# Objective
Forecast sales (no. of items sold) for each item and store for the next 28 days

20200527: Diff from m5-forecasting_v1.ipynb: Here we do predictions for 1 state at a time

20200608: Diff from m5-forecasting_v2.ipynb: Here we do hyperparam tuning

20200616: Diff from m5-forecasting_v5.ipynb: Here we remove those rows where the sell_price is NaN, as this means the item is not yet being sold, so the target value of 0 is meaningless

In [548]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import gc
import matplotlib
import numpy as np # linear algebra
import operator as op
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import time

from collections import OrderedDict
from matplotlib import pyplot as plt
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import LabelEncoder
from xgboost import XGBRegressor

%matplotlib inline

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 5GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

In [549]:
##### Input params ########
state = 'WI'
store = 'WI_3' # CA_1, CA_2, CA_3, CA_4, TX_1, TX_2, TX_3, WI_1, WI_2, WI_3

model_seed = 102
sample_seed = 103              # for sampling the dataframe

n_estimators = 100             # for the initial model before tuning. default = 100
max_depth = 3                  # for the initial model before tuning. default = 3
learning_rate = 0.1            # for the initial model before tuning. default = 0.1
min_child_weight = 1           # for the initial model before tuning. default = 1
subsample = 1                  # for the initial model before tuning. default = 1
colsample_bytree = 1           # for the initial model before tuning. default = 1
colsample_bylevel = 1          # for the initial model before tuning. default = 1
train_test_split_seed = 111    # 111
model_seed = 100

# train_start = 0
# train_end = 1884               # use date_block_num from train_start to train_end as train set
# val_date_block_num = 1885      # date_block_num for validation set
# pred_start = 1886              # do pred starting from this day
# pred_end = 1913                # do pred until and include this day

train_start = 0
train_end = 1912               # use date_block_num from train_start to train_end as train set
val_date_block_num = 1913      # date_block_num for validation set
pred_start = 1914            # do pred starting from this day
pred_end = 1941              # do pred until and include this day

shift_range = [1, 2, 3, 4, 5, 6, 7, 30, 365] # Use values from last 7 days, from 30 days ago, and from 365 days ago

fontsize = 14
ticklabelsize = 14
###########################

In [550]:
ticStart = time.time()

# Common functions

In [551]:
def downcast_dtypes(df):
    '''
        Changes column types in the dataframe: 
                
                `float64` type to `float32`
                `int64`   type to `int16`
    '''
    
    # Select columns to downcast
    float_cols = [c for c in df if df[c].dtype == "float64"]
    int_cols =   [c for c in df if df[c].dtype == "int64"]
    
    # Downcast
    df[float_cols] = df[float_cols].astype(np.float32)
    df[int_cols]   = df[int_cols].astype(np.int16)
    
    return df

def get_lags(index_cols, cols_to_rename, all_data, shift_range):
    """
    For a dataframe all_data, get the lags specified in cols_to_rename
    Input
        index_cols    : e.g. ['item_id', 'store_id', 'date_block_num']. These are the cols to do merging with the
                        lagged dataset
        cols_to_rename: e.g. ['target']. Based on this example, all_data will be returned with columns target_lag_1, 
                        target_lag_2, ..., target_lag_365.
        all_data      : need to have a column 'date_block_num' to indicate chronological order
        shift_range   : e.g. [1, 2, 3, 4, 5, 6, 7, 30, 365]
    Output
        all_data
    """
    for day_shift in shift_range:
        train_shift = all_data[index_cols + cols_to_rename].copy()
    
        # E.g. variable of 0 becomes 1, for day_shift = 1.
        # So when this is merged with variable of 1 in all_data, this will represent lag of 1.
        train_shift['date_block_num'] = train_shift['date_block_num'] + day_shift
    
        foo = lambda x: '{}_lag_{}'.format(x, day_shift) if x in cols_to_rename else x
        train_shift = train_shift.rename(columns=foo)

        all_data = pd.merge(all_data, train_shift, on=index_cols, how='left')
    
        # Fill the NaNs with 0
        for x in cols_to_rename:
            all_data[x + '_lag_' + str(day_shift)].fillna(0, inplace=True)
    
    del train_shift
    gc.collect();

    return all_data

def get_params(error_rate):
    metric1_opt = list(error_rate.items())[0][0]
    metric2_opt = list(error_rate.items())[0][1][0][0]
    mse_min = list(error_rate.items())[0][1][0][1]
    for item in error_rate:
        metric2, mse = min(error_rate[item], key=op.itemgetter(1))
        if mse < mse_min:
            metric1_opt = item
            metric2_opt = metric2
            mse_min = mse

    return (metric1_opt, metric2_opt, mse_min)


# Load data

In [552]:
sell_prices = pd.read_csv("./data/sell_prices.csv")
sell_prices.head()

store_id        item_id  wm_yr_wk  sell_price
0     CA_1  HOBBIES_1_001     11325        9.58
1     CA_1  HOBBIES_1_001     11326        9.58
2     CA_1  HOBBIES_1_001     11327        8.26
3     CA_1  HOBBIES_1_001     11328        8.26
4     CA_1  HOBBIES_1_001     11329        8.26

File 2: “sell_prices.csv”

Contains information about the price of the products sold per store and date.
* store_id: The id of the store where the product is sold. 
* item_id: The id of the product.
* wm_yr_wk: The id of the week.
* sell_price: The price of the product for the given week/store. The price is provided per week (average across seven days). If not available, this means that the product was not sold during the examined week. Note that although prices are constant at weekly basis, they may change through time (both training and test set).

In [553]:
calendar = pd.read_csv("./data/calendar.csv")
calendar

date  wm_yr_wk    weekday  wday  month  year       d  \
0     2011-01-29     11101   Saturday     1      1  2011     d_1   
1     2011-01-30     11101     Sunday     2      1  2011     d_2   
2     2011-01-31     11101     Monday     3      1  2011     d_3   
3     2011-02-01     11101    Tuesday     4      2  2011     d_4   
4     2011-02-02     11101  Wednesday     5      2  2011     d_5   
5     2011-02-03     11101   Thursday     6      2  2011     d_6   
6     2011-02-04     11101     Friday     7      2  2011     d_7   
7     2011-02-05     11102   Saturday     1      2  2011     d_8   
8     2011-02-06     11102     Sunday     2      2  2011     d_9   
9     2011-02-07     11102     Monday     3      2  2011    d_10   
10    2011-02-08     11102    Tuesday     4      2  2011    d_11   
11    2011-02-09     11102  Wednesday     5      2  2011    d_12   
12    2011-02-10     11102   Thursday     6      2  2011    d_13   
13    2011-02-11     11102     Friday     7      2  2011    d_14   
14    2011-02-12     11103   Saturday     1      2  2011    d_15   
15    2011-02-13     11103     Sunday     2      2  2011    d_16   
16    2011-02-14     11103     Monday     3      2  2011    d_17   
17    2011-02-15     11103    Tuesday     4      2  2011    d_18   
18    2011-02-16     11103  Wednesday     5      2  2011    d_19   
19    2011-02-17     11103   Thursday     6      2  2011    d_20   
20    2011-02-18     11103     Friday     7      2  2011    d_21   
21    2011-02-19     11104   Saturday     1      2  2011    d_22   
22    2011-02-20     11104     Sunday     2      2  2011    d_23   
23    2011-02-21     11104     Monday     3      2  2011    d_24   
24    2011-02-22     11104    Tuesday     4      2  2011    d_25   
25    2011-02-23     11104  Wednesday     5      2  2011    d_26   
26    2011-02-24     11104   Thursday     6      2  2011    d_27   
27    2011-02-25     11104     Friday     7      2  2011    d_28   
28    2011-02-26     11105   Saturday     1      2  2011    d_29   
29    2011-02-27     11105     Sunday     2      2  2011    d_30   
...          ...       ...        ...   ...    ...   ...     ...   
1939  2016-05-21     11617   Saturday     1      5  2016  d_1940   
1940  2016-05-22     11617     Sunday     2      5  2016  d_1941   
1941  2016-05-23     11617     Monday     3      5  2016  d_1942   
1942  2016-05-24     11617    Tuesday     4      5  2016  d_1943   
1943  2016-05-25     11617  Wednesday     5      5  2016  d_1944   
1944  2016-05-26     11617   Thursday     6      5  2016  d_1945   
1945  2016-05-27     11617     Friday     7      5  2016  d_1946   
1946  2016-05-28     11618   Saturday     1      5  2016  d_1947   
1947  2016-05-29     11618     Sunday     2      5  2016  d_1948   
1948  2016-05-30     11618     Monday     3      5  2016  d_1949   
1949  2016-05-31     11618    Tuesday     4      5  2016  d_1950   
1950  2016-06-01     11618  Wednesday     5      6  2016  d_1951   
1951  2016-06-02     11618   Thursday     6      6  2016  d_1952   
1952  2016-06-03     11618     Friday     7      6  2016  d_1953   
1953  2016-06-04     11619   Saturday     1      6  2016  d_1954   
1954  2016-06-05     11619     Sunday     2      6  2016  d_1955   
1955  2016-06-06     11619     Monday     3      6  2016  d_1956   
1956  2016-06-07     11619    Tuesday     4      6  2016  d_1957   
1957  2016-06-08     11619  Wednesday     5      6  2016  d_1958   
1958  2016-06-09     11619   Thursday     6      6  2016  d_1959   
1959  2016-06-10     11619     Friday     7      6  2016  d_1960   
1960  2016-06-11     11620   Saturday     1      6  2016  d_1961   
1961  2016-06-12     11620     Sunday     2      6  2016  d_1962   
1962  2016-06-13     11620     Monday     3      6  2016  d_1963   
1963  2016-06-14     11620    Tuesday     4      6  2016  d_1964   
1964  2016-06-15     11620  Wednesday     5      6  2016  d_1965   
1965  2016-06-16     11620   Thursday     6      6  2016  d_1966   


File 1: “calendar.csv” 

Contains information about the dates the products are sold.
* date: The date in a “y-m-d” format.
* wm_yr_wk: The id of the week the date belongs to.
* weekday: The type of the day (Saturday, Sunday, …, Friday).
* wday: The id of the weekday, starting from Saturday.
* month: The month of the date.
* year: The year of the date.
* event_name_1: If the date includes an event, the name of this event.
* event_type_1: If the date includes an event, the type of this event.
* event_name_2: If the date includes a second event, the name of this event.
* event_type_2: If the date includes a second event, the type of this event.
* snap_CA, snap_TX, and snap_WI: A binary variable (0 or 1) indicating whether the stores of CA, TX or WI allow SNAP purchases on the examined date. 1 indicates that SNAP purchases are allowed.


In [554]:
sales_tr_val = pd.read_csv("./data/sales_train_validation.csv")
print("len(sales_tr_val) = " + str(len(sales_tr_val)))
sales_tr_val.head()

len(sales_tr_val) = 30490


id        item_id    dept_id   cat_id store_id  \
0  HOBBIES_1_001_CA_1_validation  HOBBIES_1_001  HOBBIES_1  HOBBIES     CA_1   
1  HOBBIES_1_002_CA_1_validation  HOBBIES_1_002  HOBBIES_1  HOBBIES     CA_1   
2  HOBBIES_1_003_CA_1_validation  HOBBIES_1_003  HOBBIES_1  HOBBIES     CA_1   
3  HOBBIES_1_004_CA_1_validation  HOBBIES_1_004  HOBBIES_1  HOBBIES     CA_1   
4  HOBBIES_1_005_CA_1_validation  HOBBIES_1_005  HOBBIES_1  HOBBIES     CA_1   

  state_id  d_1  d_2  d_3  d_4  ...  d_1904  d_1905  d_1906  d_1907  d_1908  \
0       CA    0    0    0    0  ...       1       3       0       1       1   
1       CA    0    0    0    0  ...       0       0       0       0       0   
2       CA    0    0    0    0  ...       2       1       2       1       1   
3       CA    0    0    0    0  ...       1       0       5       4       1   
4       CA    0    0    0    0  ...       2       1       1       0       1   

   d_1909  d_1910  d_1911  d_1912  d_1913  
0       1       3       0       1       1  
1       1       0       0       0       0  
2       1       0       1       1       1  
3       0       1       3       7       2  
4       1       2       2       2       4  

[5 rows x 1919 columns]

For each (item_id, store_id), we have the number of items sold for 1913 days or 5.24 years.

There are 30,490 distinct id's, and 1913 days which gives 30,490*1913 = 58,327,370

In [555]:
# Keep only data from one state
sales_tr_val = sales_tr_val[(sales_tr_val['state_id']==state) & (sales_tr_val['store_id']==store)]
print("len(sales_tr_val) = " + str(len(sales_tr_val)))

len(sales_tr_val) = 3049


In [556]:
sample_sub = pd.read_csv("./data/sample_submission.csv")
print(len(sample_sub))
sample_sub.head()

60980


id  F1  F2  F3  F4  F5  F6  F7  F8  F9  ...  \
0  HOBBIES_1_001_CA_1_validation   0   0   0   0   0   0   0   0   0  ...   
1  HOBBIES_1_002_CA_1_validation   0   0   0   0   0   0   0   0   0  ...   
2  HOBBIES_1_003_CA_1_validation   0   0   0   0   0   0   0   0   0  ...   
3  HOBBIES_1_004_CA_1_validation   0   0   0   0   0   0   0   0   0  ...   
4  HOBBIES_1_005_CA_1_validation   0   0   0   0   0   0   0   0   0  ...   

   F19  F20  F21  F22  F23  F24  F25  F26  F27  F28  
0    0    0    0    0    0    0    0    0    0    0  
1    0    0    0    0    0    0    0    0    0    0  
2    0    0    0    0    0    0    0    0    0    0  
3    0    0    0    0    0    0    0    0    0    0  
4    0    0    0    0    0    0    0    0    0    0  

[5 rows x 29 columns]

Provide the forecast for each id for the next 28 days. For the validation rows, this corresponds to d_1914 - d_1941, and for the evaluation rows, this corresponds to d_1942 - d_1969. (Note: a month before the competition close, the ground truth for the validation rows will be provided.)

Note 30,490*2 = 60,980 which is the number of rows in the submission file.

# Pre-process data

In [557]:
print(calendar.memory_usage().sum())

# # In calendar 'd' column, change 'd_1' to 1
# calendar['d'] = calendar.apply(lambda row: row['d'].split('_')[1], axis=1)

# Replance NaN with nil
calendar = calendar.fillna(value = 'nil')

# Downcast dtypes from 64 bit to save memory
calendar= downcast_dtypes(calendar)

# Change to category dtype
calendar[["year", "event_name_1", "event_type_1", "event_name_2", "event_type_2"]] = \
    calendar[["year", "event_name_1", "event_type_1", "event_name_2", "event_type_2"]] .astype("category")

# Drop weekday column since its not used
calendar.drop(['weekday'], axis=1, inplace=True)

print(calendar.memory_usage().sum())

calendar.head()

220608
67297


date  wm_yr_wk  wday  month  year    d event_name_1 event_type_1  \
0  2011-01-29     11101     1      1  2011  d_1          nil          nil   
1  2011-01-30     11101     2      1  2011  d_2          nil          nil   
2  2011-01-31     11101     3      1  2011  d_3          nil          nil   
3  2011-02-01     11101     4      2  2011  d_4          nil          nil   
4  2011-02-02     11101     5      2  2011  d_5          nil          nil   

  event_name_2 event_type_2  snap_CA  snap_TX  snap_WI  
0          nil          nil        0        0        0  
1          nil          nil        0        0        0  
2          nil          nil        0        0        0  
3          nil          nil        1        1        0  
4          nil          nil        1        0        1

In [558]:
# Downcast dtypes from 64 bit to save memory
print(sell_prices.memory_usage().sum())
sell_prices = downcast_dtypes(sell_prices)

sell_prices[["store_id", "item_id"]] = \
    sell_prices[["store_id","item_id"]].astype("category")

print(sell_prices.memory_usage().sum())

218915952
61676881


In [559]:
# Downcast dtypes from 64 bit to save memory
print(sales_tr_val.memory_usage().sum())
sales_tr_val = downcast_dtypes(sales_tr_val)

sales_tr_val[["item_id", "dept_id", "cat_id", "store_id", "state_id"]] = \
    sales_tr_val[["item_id", "dept_id", "cat_id", "store_id", "state_id"]].astype("category")

print(sales_tr_val.memory_usage().sum())

46832640
11839520


In [560]:
# Melt sales_tr_val to switch from a wide to a long dataframe
tic = time.time()
value_vars = ['d_' + str(x) for x in range(1, 1914)]
sale_tr_val_melt = pd.melt(sales_tr_val, 
                           id_vars=['id', 'item_id', 'dept_id', 'cat_id', 'store_id', 'state_id'], 
                           value_vars=value_vars)

toc = time.time()
print("Time taken = " + str(toc-tic) + ' s')
sale_tr_val_melt.head()

Time taken = 13.031988143920898 s


id        item_id    dept_id   cat_id store_id  \
0  HOBBIES_1_001_WI_3_validation  HOBBIES_1_001  HOBBIES_1  HOBBIES     WI_3   
1  HOBBIES_1_002_WI_3_validation  HOBBIES_1_002  HOBBIES_1  HOBBIES     WI_3   
2  HOBBIES_1_003_WI_3_validation  HOBBIES_1_003  HOBBIES_1  HOBBIES     WI_3   
3  HOBBIES_1_004_WI_3_validation  HOBBIES_1_004  HOBBIES_1  HOBBIES     WI_3   
4  HOBBIES_1_005_WI_3_validation  HOBBIES_1_005  HOBBIES_1  HOBBIES     WI_3   

  state_id variable  value  
0       WI      d_1      0  
1       WI      d_1      0  
2       WI      d_1      0  
3       WI      d_1      4  
4       WI      d_1      0

In [561]:
del sales_tr_val
gc.collect();
sales_tr_val = pd.DataFrame()

In [562]:
# Merge with calendar to get events info
sale_tr_val_melt_merged = sale_tr_val_melt.merge(calendar, 
                                                 left_on='variable', 
                                                 right_on='d', 
                                                 how='left')

# Drop unneccesary columns
sale_tr_val_melt_merged.drop(columns=['d'], inplace=True)

sale_tr_val_melt_merged.head()

id        item_id    dept_id   cat_id store_id  \
0  HOBBIES_1_001_WI_3_validation  HOBBIES_1_001  HOBBIES_1  HOBBIES     WI_3   
1  HOBBIES_1_002_WI_3_validation  HOBBIES_1_002  HOBBIES_1  HOBBIES     WI_3   
2  HOBBIES_1_003_WI_3_validation  HOBBIES_1_003  HOBBIES_1  HOBBIES     WI_3   
3  HOBBIES_1_004_WI_3_validation  HOBBIES_1_004  HOBBIES_1  HOBBIES     WI_3   
4  HOBBIES_1_005_WI_3_validation  HOBBIES_1_005  HOBBIES_1  HOBBIES     WI_3   

  state_id variable  value        date  wm_yr_wk  wday  month  year  \
0       WI      d_1      0  2011-01-29     11101     1      1  2011   
1       WI      d_1      0  2011-01-29     11101     1      1  2011   
2       WI      d_1      0  2011-01-29     11101     1      1  2011   
3       WI      d_1      4  2011-01-29     11101     1      1  2011   
4       WI      d_1      0  2011-01-29     11101     1      1  2011   

  event_name_1 event_type_1 event_name_2 event_type_2  snap_CA  snap_TX  \
0          nil          nil          nil          nil        0        0   
1          nil          nil          nil          nil        0        0   
2          nil          nil          nil          nil        0        0   
3          nil          nil          nil          nil        0        0   
4          nil          nil          nil          nil        0        0   

   snap_WI  
0        0  
1        0  
2        0  
3        0  
4        0

In [563]:
# Merge with sell_prices to get price info
sale_tr_val_melt_merged2 = sale_tr_val_melt_merged.merge(sell_prices, 
                                                         left_on=['store_id', 'item_id', 'wm_yr_wk'], 
                                                         right_on=['store_id', 'item_id', 'wm_yr_wk'], 
                                                         how='left')
sale_tr_val_melt_merged2

id        item_id    dept_id   cat_id  \
0        HOBBIES_1_001_WI_3_validation  HOBBIES_1_001  HOBBIES_1  HOBBIES   
1        HOBBIES_1_002_WI_3_validation  HOBBIES_1_002  HOBBIES_1  HOBBIES   
2        HOBBIES_1_003_WI_3_validation  HOBBIES_1_003  HOBBIES_1  HOBBIES   
3        HOBBIES_1_004_WI_3_validation  HOBBIES_1_004  HOBBIES_1  HOBBIES   
4        HOBBIES_1_005_WI_3_validation  HOBBIES_1_005  HOBBIES_1  HOBBIES   
5        HOBBIES_1_006_WI_3_validation  HOBBIES_1_006  HOBBIES_1  HOBBIES   
6        HOBBIES_1_007_WI_3_validation  HOBBIES_1_007  HOBBIES_1  HOBBIES   
7        HOBBIES_1_008_WI_3_validation  HOBBIES_1_008  HOBBIES_1  HOBBIES   
8        HOBBIES_1_009_WI_3_validation  HOBBIES_1_009  HOBBIES_1  HOBBIES   
9        HOBBIES_1_010_WI_3_validation  HOBBIES_1_010  HOBBIES_1  HOBBIES   
10       HOBBIES_1_011_WI_3_validation  HOBBIES_1_011  HOBBIES_1  HOBBIES   
11       HOBBIES_1_012_WI_3_validation  HOBBIES_1_012  HOBBIES_1  HOBBIES   
12       HOBBIES_1_013_WI_3_validation  HOBBIES_1_013  HOBBIES_1  HOBBIES   
13       HOBBIES_1_014_WI_3_validation  HOBBIES_1_014  HOBBIES_1  HOBBIES   
14       HOBBIES_1_015_WI_3_validation  HOBBIES_1_015  HOBBIES_1  HOBBIES   
15       HOBBIES_1_016_WI_3_validation  HOBBIES_1_016  HOBBIES_1  HOBBIES   
16       HOBBIES_1_017_WI_3_validation  HOBBIES_1_017  HOBBIES_1  HOBBIES   
17       HOBBIES_1_018_WI_3_validation  HOBBIES_1_018  HOBBIES_1  HOBBIES   
18       HOBBIES_1_019_WI_3_validation  HOBBIES_1_019  HOBBIES_1  HOBBIES   
19       HOBBIES_1_020_WI_3_validation  HOBBIES_1_020  HOBBIES_1  HOBBIES   
20       HOBBIES_1_021_WI_3_validation  HOBBIES_1_021  HOBBIES_1  HOBBIES   
21       HOBBIES_1_022_WI_3_validation  HOBBIES_1_022  HOBBIES_1  HOBBIES   
22       HOBBIES_1_023_WI_3_validation  HOBBIES_1_023  HOBBIES_1  HOBBIES   
23       HOBBIES_1_024_WI_3_validation  HOBBIES_1_024  HOBBIES_1  HOBBIES   
24       HOBBIES_1_025_WI_3_validation  HOBBIES_1_025  HOBBIES_1  HOBBIES   
25       HOBBIES_1_026_WI_3_validation  HOBBIES_1_026  HOBBIES_1  HOBBIES   
26       HOBBIES_1_027_WI_3_validation  HOBBIES_1_027  HOBBIES_1  HOBBIES   
27       HOBBIES_1_028_WI_3_validation  HOBBIES_1_028  HOBBIES_1  HOBBIES   
28       HOBBIES_1_029_WI_3_validation  HOBBIES_1_029  HOBBIES_1  HOBBIES   
29       HOBBIES_1_030_WI_3_validation  HOBBIES_1_030  HOBBIES_1  HOBBIES   
...                                ...            ...        ...      ...   
5832707    FOODS_3_798_WI_3_validation    FOODS_3_798    FOODS_3    FOODS   
5832708    FOODS_3_799_WI_3_validation    FOODS_3_799    FOODS_3    FOODS   
5832709    FOODS_3_800_WI_3_validation    FOODS_3_800    FOODS_3    FOODS   
5832710    FOODS_3_801_WI_3_validation    FOODS_3_801    FOODS_3    FOODS   
5832711    FOODS_3_802_WI_3_validation    FOODS_3_802    FOODS_3    FOODS   
5832712    FOODS_3_803_WI_3_validation    FOODS_3_803    FOODS_3    FOODS   
5832713    FOODS_3_804_WI_3_validation    FOODS_3_804    FOODS_3    FOODS   
5832714    FOODS_3_805_WI_3_validation    FOODS_3_805    FOODS_3    FOODS   
5832715    FOODS_3_806_WI_3_validation    FOODS_3_806    FOODS_3    FOODS   
5832716    FOODS_3_807_WI_3_validation    FOODS_3_807    FOODS_3    FOODS   
5832717    FOODS_3_808_WI_3_validation    FOODS_3_808    FOODS_3    FOODS   
5832718    FOODS_3_809_WI_3_validation    FOODS_3_809    FOODS_3    FOODS   
5832719    FOODS_3_810_WI_3_validation    FOODS_3_810    FOODS_3    FOODS   
5832720    FOODS_3_811_WI_3_validation    FOODS_3_811    FOODS_3    FOODS   
5832721    FOODS_3_812_WI_3_validation    FOODS_3_812    FOODS_3    FOODS   
5832722    FOODS_3_813_WI_3_validation    FOODS_3_813    FOODS_3    FOODS   
5832723    FOODS_3_814_WI_3_validation    FOODS_3_814    FOODS_3    FOODS   
5832724    FOODS_3_815_WI_3_validation    FOODS_3_815    FOODS_3    FOODS   
5832725    FOODS_3_816_WI_3_validation    FOODS_3_816    FOODS_3    FOODS   
5832726    FOODS_3_817_WI_3_validation    FOODS_3_817    FOODS_3    FOODS   
5832727    FOODS_3_818_WI_3_vali

In [564]:
# Check whether there are rows where value != 0 and sell_price is null
sale_tr_val_melt_merged2[(sale_tr_val_melt_merged2['value']!=0) & (sale_tr_val_melt_merged2['sell_price'].isnull())]

Empty DataFrame
Columns: [id, item_id, dept_id, cat_id, store_id, state_id, variable, value, date, wm_yr_wk, wday, month, year, event_name_1, event_type_1, event_name_2, event_type_2, snap_CA, snap_TX, snap_WI, sell_price]
Index: []

[0 rows x 21 columns]

There are no rows where value != 0 and sell_price is null

In [565]:
# Remove those rows where sell price is NaN and value=0 as this means the item is not being sold yet
print("Before dropna, len(sale_tr_val_melt_merged2) = " + str(len(sale_tr_val_melt_merged2)))
sale_tr_val_melt_merged2.dropna(axis=0, how='any', subset=['sell_price'], inplace=True)
print("After dropna, len(sale_tr_val_melt_merged2) = " + str(len(sale_tr_val_melt_merged2)))

Before dropna, len(sale_tr_val_melt_merged2) = 5832737
After dropna, len(sale_tr_val_melt_merged2) = 4686669


In [566]:
# del sell_prices
# gc.collect();
# sell_prices = pd.DataFrame()

del sale_tr_val_melt_merged
gc.collect();
sale_tr_val_melt_merged = pd.DataFrame()

In [567]:
# Get no. of distinct values in each column
print("item_id uniques = " + str(sale_tr_val_melt_merged2['item_id'].nunique()))
print("dept_id uniques = " + str(sale_tr_val_melt_merged2['dept_id'].nunique()))
print("cat_id uniques = " + str(sale_tr_val_melt_merged2['cat_id'].nunique()))
print("store_id uniques = " + str(sale_tr_val_melt_merged2['store_id'].nunique()))
print("state_id uniques = " + str(sale_tr_val_melt_merged2['state_id'].nunique()))

item_id uniques = 3049
dept_id uniques = 7
cat_id uniques = 3
store_id uniques = 1
state_id uniques = 1


In [568]:
print(sale_tr_val_melt_merged2['dept_id'].unique())

[HOBBIES_1, HOBBIES_2, HOUSEHOLD_1, HOUSEHOLD_2, FOODS_1, FOODS_2, FOODS_3]
Categories (7, object): [HOBBIES_1, HOBBIES_2, HOUSEHOLD_1, HOUSEHOLD_2, FOODS_1, FOODS_2, FOODS_3]


In [569]:
print(sale_tr_val_melt_merged2['cat_id'].unique())

[HOBBIES, HOUSEHOLD, FOODS]
Categories (3, object): [HOBBIES, HOUSEHOLD, FOODS]


In [570]:
print(sale_tr_val_melt_merged2['store_id'].unique())

['WI_3']


In [571]:
print(sale_tr_val_melt_merged2['state_id'].unique())

[WI]
Categories (1, object): [WI]


In [572]:
# Convert variable to a numerical column ie. d_1 to 1
sale_tr_val_melt_merged2['variable'] = sale_tr_val_melt_merged2['variable'].str.split('_').str[1]
sale_tr_val_melt_merged2

id        item_id    dept_id   cat_id  \
3        HOBBIES_1_004_WI_3_validation  HOBBIES_1_004  HOBBIES_1  HOBBIES   
7        HOBBIES_1_008_WI_3_validation  HOBBIES_1_008  HOBBIES_1  HOBBIES   
8        HOBBIES_1_009_WI_3_validation  HOBBIES_1_009  HOBBIES_1  HOBBIES   
9        HOBBIES_1_010_WI_3_validation  HOBBIES_1_010  HOBBIES_1  HOBBIES   
11       HOBBIES_1_012_WI_3_validation  HOBBIES_1_012  HOBBIES_1  HOBBIES   
14       HOBBIES_1_015_WI_3_validation  HOBBIES_1_015  HOBBIES_1  HOBBIES   
15       HOBBIES_1_016_WI_3_validation  HOBBIES_1_016  HOBBIES_1  HOBBIES   
16       HOBBIES_1_017_WI_3_validation  HOBBIES_1_017  HOBBIES_1  HOBBIES   
19       HOBBIES_1_020_WI_3_validation  HOBBIES_1_020  HOBBIES_1  HOBBIES   
21       HOBBIES_1_022_WI_3_validation  HOBBIES_1_022  HOBBIES_1  HOBBIES   
22       HOBBIES_1_023_WI_3_validation  HOBBIES_1_023  HOBBIES_1  HOBBIES   
24       HOBBIES_1_025_WI_3_validation  HOBBIES_1_025  HOBBIES_1  HOBBIES   
27       HOBBIES_1_028_WI_3_validation  HOBBIES_1_028  HOBBIES_1  HOBBIES   
28       HOBBIES_1_029_WI_3_validation  HOBBIES_1_029  HOBBIES_1  HOBBIES   
30       HOBBIES_1_031_WI_3_validation  HOBBIES_1_031  HOBBIES_1  HOBBIES   
31       HOBBIES_1_032_WI_3_validation  HOBBIES_1_032  HOBBIES_1  HOBBIES   
35       HOBBIES_1_036_WI_3_validation  HOBBIES_1_036  HOBBIES_1  HOBBIES   
37       HOBBIES_1_038_WI_3_validation  HOBBIES_1_038  HOBBIES_1  HOBBIES   
42       HOBBIES_1_043_WI_3_validation  HOBBIES_1_043  HOBBIES_1  HOBBIES   
43       HOBBIES_1_044_WI_3_validation  HOBBIES_1_044  HOBBIES_1  HOBBIES   
46       HOBBIES_1_047_WI_3_validation  HOBBIES_1_047  HOBBIES_1  HOBBIES   
49       HOBBIES_1_050_WI_3_validation  HOBBIES_1_050  HOBBIES_1  HOBBIES   
54       HOBBIES_1_055_WI_3_validation  HOBBIES_1_055  HOBBIES_1  HOBBIES   
57       HOBBIES_1_058_WI_3_validation  HOBBIES_1_058  HOBBIES_1  HOBBIES   
61       HOBBIES_1_063_WI_3_validation  HOBBIES_1_063  HOBBIES_1  HOBBIES   
62       HOBBIES_1_064_WI_3_validation  HOBBIES_1_064  HOBBIES_1  HOBBIES   
65       HOBBIES_1_067_WI_3_validation  HOBBIES_1_067  HOBBIES_1  HOBBIES   
66       HOBBIES_1_068_WI_3_validation  HOBBIES_1_068  HOBBIES_1  HOBBIES   
70       HOBBIES_1_073_WI_3_validation  HOBBIES_1_073  HOBBIES_1  HOBBIES   
71       HOBBIES_1_074_WI_3_validation  HOBBIES_1_074  HOBBIES_1  HOBBIES   
...                                ...            ...        ...      ...   
5832707    FOODS_3_798_WI_3_validation    FOODS_3_798    FOODS_3    FOODS   
5832708    FOODS_3_799_WI_3_validation    FOODS_3_799    FOODS_3    FOODS   
5832709    FOODS_3_800_WI_3_validation    FOODS_3_800    FOODS_3    FOODS   
5832710    FOODS_3_801_WI_3_validation    FOODS_3_801    FOODS_3    FOODS   
5832711    FOODS_3_802_WI_3_validation    FOODS_3_802    FOODS_3    FOODS   
5832712    FOODS_3_803_WI_3_validation    FOODS_3_803    FOODS_3    FOODS   
5832713    FOODS_3_804_WI_3_validation    FOODS_3_804    FOODS_3    FOODS   
5832714    FOODS_3_805_WI_3_validation    FOODS_3_805    FOODS_3    FOODS   
5832715    FOODS_3_806_WI_3_validation    FOODS_3_806    FOODS_3    FOODS   
5832716    FOODS_3_807_WI_3_validation    FOODS_3_807    FOODS_3    FOODS   
5832717    FOODS_3_808_WI_3_validation    FOODS_3_808    FOODS_3    FOODS   
5832718    FOODS_3_809_WI_3_validation    FOODS_3_809    FOODS_3    FOODS   
5832719    FOODS_3_810_WI_3_validation    FOODS_3_810    FOODS_3    FOODS   
5832720    FOODS_3_811_WI_3_validation    FOODS_3_811    FOODS_3    FOODS   
5832721    FOODS_3_812_WI_3_validation    FOODS_3_812    FOODS_3    FOODS   
5832722    FOODS_3_813_WI_3_validation    FOODS_3_813    FOODS_3    FOODS   
5832723    FOODS_3_814_WI_3_validation    FOODS_3_814    FOODS_3    FOODS   
5832724    FOODS_3_815_WI_3_validation    FOODS_3_815    FOODS_3    FOODS   
5832725    FOODS_3_816_WI_3_validation    FOODS_3_816    FOODS_3    FOODS   
5832726    FOODS_3_817_WI_3_validation    FOODS_3_817    FOODS_3    FOODS   
5832727    FOODS_3_818_WI_3_vali

In [573]:
# Convert variable to int
sale_tr_val_melt_merged2['variable'] = sale_tr_val_melt_merged2['variable'].astype('int16')

# Create features

In [574]:
all_data = sale_tr_val_melt_merged2
all_data = all_data.rename(columns={'variable': 'date_block_num', 
                                    'value': 'target'})

In [575]:
# Remove unnecessaray column
all_data.drop(['date'], axis=1, inplace=True)

In [576]:
# Convert features to categorical
columns_to_be_labeled = [
    'year',
    'event_name_1',
    'event_type_1',
    'event_name_2',
    'event_type_2',
    'wm_yr_wk',
    'item_id',
    'store_id',
    'dept_id',
    'cat_id',
    'state_id'
]
for column in columns_to_be_labeled:
    all_data.loc[:, column] = LabelEncoder().fit_transform(all_data[column]) # fit_transform: Fit label encoder and return encoded labels. Note the encoded labels are integers!!

all_data.head()

id  item_id  dept_id  cat_id  store_id  \
3   HOBBIES_1_004_WI_3_validation     1440        3       1         0   
7   HOBBIES_1_008_WI_3_validation     1444        3       1         0   
8   HOBBIES_1_009_WI_3_validation     1445        3       1         0   
9   HOBBIES_1_010_WI_3_validation     1446        3       1         0   
11  HOBBIES_1_012_WI_3_validation     1448        3       1         0   

    state_id  date_block_num  target  wm_yr_wk  wday  month  year  \
3          0               1       4         0     1      1     0   
7          0               1       6         0     1      1     0   
8          0               1       1         0     1      1     0   
9          0               1       0         0     1      1     0   
11         0               1       3         0     1      1     0   

    event_name_1  event_type_1  event_name_2  event_type_2  snap_CA  snap_TX  \
3             30             4             4             2        0        0   
7             30             4             4             2        0        0   
8             30             4             4             2        0        0   
9             30             4             4             2        0        0   
11            30             4             4             2        0        0   

    snap_WI  sell_price  
3         0        4.34  
7         0        0.46  
8         0        1.56  
9         0        3.17  
11        0        5.98

In [577]:
index_cols = ['item_id', 'store_id', 'date_block_num']

cols_to_rename = ['target']

for day_shift in shift_range:
    print(day_shift)
    train_shift = all_data[index_cols + cols_to_rename].copy()
    
    # E.g. variable of 0 becomes 1, for day_shift = 1.
    # So when this is merged with variable of 1 in all_data, this will represent lag of 1.
    train_shift['date_block_num'] = train_shift['date_block_num'] + day_shift
    
    foo = lambda x: '{}_lag_{}'.format(x, day_shift) if x in cols_to_rename else x
    train_shift = train_shift.rename(columns=foo)

    all_data = pd.merge(all_data, train_shift, on=index_cols, how='left')
    
#     # Fill the NaNs with 0
#     for x in cols_to_rename:
#         all_data[x + '_lag_' + str(day_shift)].fillna(0, inplace=True)
    
del train_shift
gc.collect();

all_data

1
2
3
4
5
6
7
30
365


id  item_id  dept_id  cat_id  store_id  \
0        HOBBIES_1_004_WI_3_validation     1440        3       1         0   
1        HOBBIES_1_008_WI_3_validation     1444        3       1         0   
2        HOBBIES_1_009_WI_3_validation     1445        3       1         0   
3        HOBBIES_1_010_WI_3_validation     1446        3       1         0   
4        HOBBIES_1_012_WI_3_validation     1448        3       1         0   
5        HOBBIES_1_015_WI_3_validation     1451        3       1         0   
6        HOBBIES_1_016_WI_3_validation     1452        3       1         0   
7        HOBBIES_1_017_WI_3_validation     1453        3       1         0   
8        HOBBIES_1_020_WI_3_validation     1456        3       1         0   
9        HOBBIES_1_022_WI_3_validation     1458        3       1         0   
10       HOBBIES_1_023_WI_3_validation     1459        3       1         0   
11       HOBBIES_1_025_WI_3_validation     1461        3       1         0   
12       HOBBIES_1_028_WI_3_validation     1464        3       1         0   
13       HOBBIES_1_029_WI_3_validation     1465        3       1         0   
14       HOBBIES_1_031_WI_3_validation     1467        3       1         0   
15       HOBBIES_1_032_WI_3_validation     1468        3       1         0   
16       HOBBIES_1_036_WI_3_validation     1472        3       1         0   
17       HOBBIES_1_038_WI_3_validation     1474        3       1         0   
18       HOBBIES_1_043_WI_3_validation     1479        3       1         0   
19       HOBBIES_1_044_WI_3_validation     1480        3       1         0   
20       HOBBIES_1_047_WI_3_validation     1483        3       1         0   
21       HOBBIES_1_050_WI_3_validation     1486        3       1         0   
22       HOBBIES_1_055_WI_3_validation     1491        3       1         0   
23       HOBBIES_1_058_WI_3_validation     1494        3       1         0   
24       HOBBIES_1_063_WI_3_validation     1498        3       1         0   
25       HOBBIES_1_064_WI_3_validation     1499        3       1         0   
26       HOBBIES_1_067_WI_3_validation     1502        3       1         0   
27       HOBBIES_1_068_WI_3_validation     1503        3       1         0   
28       HOBBIES_1_073_WI_3_validation     1507        3       1         0   
29       HOBBIES_1_074_WI_3_validation     1508        3       1         0   
...                                ...      ...      ...     ...       ...   
4686639    FOODS_3_798_WI_3_validation     1407        2       0         0   
4686640    FOODS_3_799_WI_3_validation     1408        2       0         0   
4686641    FOODS_3_800_WI_3_validation     1409        2       0         0   
4686642    FOODS_3_801_WI_3_validation     1410        2       0         0   
4686643    FOODS_3_802_WI_3_validation     1411        2       0         0   
4686644    FOODS_3_803_WI_3_validation     1412        2       0         0   
4686645    FOODS_3_804_WI_3_validation     1413        2       0         0   
4686646    FOODS_3_805_WI_3_validation     1414        2       0         0   
4686647    FOODS_3_806_WI_3_validation     1415        2       0         0   
4686648    FOODS_3_807_WI_3_validation     1416        2       0         0   
4686649    FOODS_3_808_WI_3_validation     1417        2       0         0   
4686650    FOODS_3_809_WI_3_validation     1418        2       0         0   
4686651    FOODS_3_810_WI_3_validation     1419        2       0         0   
4686652    FOODS_3_811_WI_3_validation     1420        2       0         0   
4686653    FOODS_3_812_WI_3_validation     1421        2       0         0   
4686654    FOODS_3_813_WI_3_validation     1422        2       0         0   
4686655    FOODS_3_814_WI_3_validation     1423        2       0         0   
4686656    FOODS_3_815_WI_3_validation     1424        2       0         0   
4686657    FOODS_3_816_WI_3_validation     1425        2       0         0   
4686658    FOODS_3_817_WI_3_validation     1426        2  

In [578]:
# Remove those rows where target_lag_x is NaN
lag_cols = []
for day_shift in shift_range[:-1]:         # ignore NaN for target_lag_365, else too many rows will be removed
    lag_cols.append('target_lag_'+str(day_shift))

print("Before dropna, len(all_data) = " + str(len(all_data)))
all_data.dropna(axis=0, how='any', subset=lag_cols, inplace=True)
print("After dropna, len(all_data) = " + str(len(all_data)))

Before dropna, len(all_data) = 4686669
After dropna, len(all_data) = 4595199


In [579]:
all_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4595199 entries, 38412 to 4686668
Data columns (total 29 columns):
id                object
item_id           int64
dept_id           int64
cat_id            int64
store_id          int64
state_id          int64
date_block_num    int16
target            int16
wm_yr_wk          int64
wday              int16
month             int16
year              int64
event_name_1      int64
event_type_1      int64
event_name_2      int64
event_type_2      int64
snap_CA           int16
snap_TX           int16
snap_WI           int16
sell_price        float32
target_lag_1      float64
target_lag_2      float64
target_lag_3      float64
target_lag_4      float64
target_lag_5      float64
target_lag_6      float64
target_lag_7      float64
target_lag_30     float64
target_lag_365    float64
dtypes: float32(1), float64(9), int16(7), int64(11), object(1)
memory usage: 850.2+ MB


# Split into train, val, test sets

We use 30 days as validation data

In [580]:
to_drop_cols = ['id', 'target', 'date_block_num', 'store_id', 'state_id']

In [581]:
X_train = all_data[(all_data['date_block_num']>=train_start) & 
                   (all_data['date_block_num']<=train_end)].drop(to_drop_cols, axis = 1)
X_cv = all_data[all_data['date_block_num']==val_date_block_num].drop(to_drop_cols, axis = 1)
X_train_cv = all_data[(all_data['date_block_num']>=train_start) & 
                      (all_data['date_block_num']<=val_date_block_num)].drop(to_drop_cols, axis = 1)

y_train = all_data[(all_data['date_block_num']>=train_start) & 
                   (all_data['date_block_num']<=train_end)][['id', 'target']]
y_cv = all_data[all_data['date_block_num']==val_date_block_num][['id', 'target']]
y_train_cv = all_data[(all_data['date_block_num']>=train_start) &
                      (all_data['date_block_num']<=val_date_block_num)][['id', 'target']]

# Train the model and get prediction for day val_date_block_num

In [582]:
# Create the model
model = XGBRegressor(seed=model_seed,
                      n_estimators=n_estimators,
                      max_depth=max_depth,
                      learning_rate=learning_rate,
                      min_child_weight=min_child_weight)

# Train the regressor
model.fit(X_train, y_train['target'])


[09:08:43] WARNING: /Users/travis/build/dmlc/xgboost/src/gbm/gbtree.cc:139: Tree method is automatically selected to be 'approx' for faster speed. To use old behavior (exact greedy algorithm on single machine), set tree_method to 'exact'.


XGBRegressor(base_score=0.5, booster='gbtree', colsample_bylevel=1,
             colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
             importance_type='gain', interaction_constraints='',
             learning_rate=0.1, max_delta_step=0, max_depth=3,
             min_child_weight=1, missing=nan, monotone_constraints='()',
             n_estimators=100, n_jobs=0, num_parallel_tree=1,
             objective='reg:squarederror', random_state=100, reg_alpha=0,
             reg_lambda=1, scale_pos_weight=1, seed=100, subsample=1,
             tree_method='approx', validate_parameters=1, verbosity=None)

In [583]:
# View a list of the features and their importance scores
imp = list(zip(X_train.columns, model.feature_importances_))
imp.sort(key=lambda tup: tup[1], reverse=False) 
imp

[('cat_id', 0.0),
 ('year', 0.0),
 ('event_name_2', 0.0),
 ('event_type_2', 0.0),
 ('snap_TX', 0.0),
 ('snap_CA', 0.0005112946),
 ('event_type_1', 0.0006217556),
 ('event_name_1', 0.0006976856),
 ('wm_yr_wk', 0.0021909513),
 ('dept_id', 0.0023557765),
 ('month', 0.0034413165),
 ('sell_price', 0.0065370942),
 ('target_lag_365', 0.0066472148),
 ('wday', 0.008295924),
 ('snap_WI', 0.008683728),
 ('item_id', 0.014257864),
 ('target_lag_30', 0.025488589),
 ('target_lag_4', 0.052244473),
 ('target_lag_5', 0.06401203),
 ('target_lag_6', 0.0997067),
 ('target_lag_2', 0.11718076),
 ('target_lag_3', 0.1432134),
 ('target_lag_7', 0.15836455),
 ('target_lag_1', 0.2855489)]

In [584]:
# Do prediction on train set
pred = model.predict(X_train).astype('int')

# Calculate RMSE
print("Train set MSE = " + str(mean_squared_error(y_train['target'], pred)))

# Do prediction on val set
pred = model.predict(X_cv).astype('int')

# Calculate RMSE
print("Val set MSE = " + str(mean_squared_error(y_cv['target'], pred)))

Train set MSE = 5.572780505863267
Val set MSE = 3.438504427681207


In [585]:
y_cv['pred'] = pred
y_cv

id  target  pred
4683620  HOBBIES_1_001_WI_3_validation       0     0
4683621  HOBBIES_1_002_WI_3_validation       0     0
4683622  HOBBIES_1_003_WI_3_validation       1     0
4683623  HOBBIES_1_004_WI_3_validation       2     0
4683624  HOBBIES_1_005_WI_3_validation       0     0
4683625  HOBBIES_1_006_WI_3_validation       2     0
4683626  HOBBIES_1_007_WI_3_validation       0     0
4683627  HOBBIES_1_008_WI_3_validation       3     2
4683628  HOBBIES_1_009_WI_3_validation       0     0
4683629  HOBBIES_1_010_WI_3_validation       0     0
4683630  HOBBIES_1_011_WI_3_validation       0     0
4683631  HOBBIES_1_012_WI_3_validation       0     0
4683632  HOBBIES_1_013_WI_3_validation       1     0
4683633  HOBBIES_1_014_WI_3_validation       0     2
4683634  HOBBIES_1_015_WI_3_validation       0     2
4683635  HOBBIES_1_016_WI_3_validation       1     2
4683636  HOBBIES_1_017_WI_3_validation       2     0
4683637  HOBBIES_1_018_WI_3_validation       0     0
4683638  HOBBIES_1_019_WI_3_validation       1     0
4683639  HOBBIES_1_020_WI_3_validation       0     0
4683640  HOBBIES_1_021_WI_3_validation       0     0
4683641  HOBBIES_1_022_WI_3_validation       0     0
4683642  HOBBIES_1_023_WI_3_validation       1     0
4683643  HOBBIES_1_024_WI_3_validation       0     0
4683644  HOBBIES_1_025_WI_3_validation       0     0
4683645  HOBBIES_1_026_WI_3_validation       2     0
4683646  HOBBIES_1_027_WI_3_validation       0     0
4683647  HOBBIES_1_028_WI_3_validation       1     0
4683648  HOBBIES_1_029_WI_3_validation       0     1
4683649  HOBBIES_1_030_WI_3_validation       0     1
...                                ...     ...   ...
4686639    FOODS_3_798_WI_3_validation       2     0
4686640    FOODS_3_799_WI_3_validation       0     0
4686641    FOODS_3_800_WI_3_validation      10     7
4686642    FOODS_3_801_WI_3_validation       2     3
4686643    FOODS_3_802_WI_3_validation       0     0
4686644    FOODS_3_803_WI_3_validation       0     0
4686645    FOODS_3_804_WI_3_validation       2     5
4686646    FOODS_3_805_WI_3_validation       1     2
4686647    FOODS_3_806_WI_3_validation       1     0
4686648    FOODS_3_807_WI_3_validation       0     3
4686649    FOODS_3_808_WI_3_validation       0     0
4686650    FOODS_3_809_WI_3_validation       0     2
4686651    FOODS_3_810_WI_3_validation       1     0
4686652    FOODS_3_811_WI_3_validation      28    14
4686653    FOODS_3_812_WI_3_validation       3     0
4686654    FOODS_3_813_WI_3_validation       0     0
4686655    FOODS_3_814_WI_3_validation       4     2
4686656    FOODS_3_815_WI_3_validation       0     0
4686657    FOODS_3_816_WI_3_validation       0     0
4686658    FOODS_3_817_WI_3_validation       1     0
4686659    FOODS_3_818_WI_3_validation       1     2
4686660    FOODS_3_819_WI_3_validation       2     1
4686661    FOODS_3_820_WI_3_validation       1     2
4686662    FOODS_3_821_WI_3_validation       0     1
4686663    FOODS_3_822_WI_3_validation       1     1
4686664    FOODS_3_823_WI_3_validation       1     0
4686665    FOODS_3_824_WI_3_validation       0     0
4686666    FOODS_3_825_WI_3_validation       0     0
4686667    FOODS_3_826_WI_3_validation       3     0
4686668    FOODS_3_827_WI_3_validation       0     0

[3049 rows x 3 columns]

# Hyperparameter tuning

## Plot the error rate versus n_estimators, varying max_depth (size of tree, default = 3)

In [586]:
# ensemble_clfs = [
#     (3,
#         XGBRegressor(seed=model_seed,
#                       max_depth=3,
#                       learning_rate=learning_rate,
#                       min_child_weight=min_child_weight)),
#     (5,
#         XGBRegressor(seed=model_seed,
#                       max_depth=5,
#                       learning_rate=learning_rate,
#                       min_child_weight=min_child_weight)),
#     (7,
#         XGBRegressor(seed=model_seed,
#                       max_depth=7,
#                       learning_rate=learning_rate,
#                       min_child_weight=min_child_weight)),
#     (9,
#         XGBRegressor(seed=model_seed,
#                       max_depth=9,
#                       learning_rate=learning_rate,
#                       min_child_weight=min_child_weight))
# ]

# # Map a classifier name to a list of (<n_estimators>, <error rate>) pairs.
# error_rate = OrderedDict((label, []) for label, _ in ensemble_clfs)

# # Range of `n_estimators` values to explore.
# estimators_list = [10]
# estimators_list.extend(range(100, 500, 200))

# tic = time.time()
# for label, clf in ensemble_clfs:
#     print(label)
#     for i in estimators_list:
#         print(i)
#         tic1 = time.time()
#         clf.set_params(n_estimators=i)
#         clf.fit(X_train, y_train['target'])
        
#         pred = clf.predict(X_cv).astype('int')
#         error_rate[label].append((i, mean_squared_error(y_cv['target'], pred)))
#         toc1 = time.time()
#         print("Time taken = " + str((toc1-tic1)/60.0) + " mins")
        
# toc = time.time()
# print("Minutes taken = " + str((toc-tic)/60.0))

In [587]:
# # Generate the "error rate" vs. "n_estimators" plot.
# plt.figure(figsize=(12, 8), dpi=80)
# for label, clf_err in error_rate.items():
#     xs, ys = zip(*clf_err)
#     plt.plot(xs, ys, label=label, marker='x')

# plt.xlabel("n_estimators")
# plt.ylabel("MSE")
# plt.legend(loc="lower right")
# plt.grid()
# matplotlib.rcParams.update({'font.size': fontsize})
# matplotlib.rcParams['xtick.labelsize'] = ticklabelsize
# matplotlib.rcParams['ytick.labelsize'] = ticklabelsize

In [588]:
# error_rate

In [589]:
# max_depth_opt, n_estimators_opt, mse_min = get_params(error_rate)
# max_depth_opt, n_estimators_opt, mse_min

## Plot the error rate versus n_estimators, varying learning_rate (default = 0.1)

In [590]:
# ensemble_clfs = [
#     (0.1,
#         XGBRegressor(seed=model_seed,
#                       max_depth=max_depth_opt,
#                       learning_rate=0.1,
#                       min_child_weight=min_child_weight)),
#     (0.01,
#         XGBRegressor(seed=model_seed,
#                       max_depth=max_depth_opt,
#                       learning_rate=0.01,
#                       min_child_weight=min_child_weight)),
#     (0.001,
#         XGBRegressor(seed=model_seed,
#                       max_depth=max_depth_opt,
#                       learning_rate=0.001,
#                       min_child_weight=min_child_weight))
# ]

# # Map a classifier name to a list of (<n_estimators>, <error rate>) pairs.
# error_rate = OrderedDict((label, []) for label, _ in ensemble_clfs)

# # Range of `n_estimators` values to explore.
# estimators_list = [10]
# estimators_list.extend(range(100, 500, 200))

# tic = time.time()
# for label, clf in ensemble_clfs:
#     print(label)
#     for i in estimators_list:
#         print(i)
#         tic1 = time.time()
#         clf.set_params(n_estimators=i)
#         clf.fit(X_train, y_train['target'])
        
#         pred = clf.predict(X_cv).astype('int')
#         error_rate[label].append((i, mean_squared_error(y_cv['target'], pred)))
#         toc1 = time.time()
#         print("Time taken = " + str((toc1-tic1)/60.0) + " mins")

# toc = time.time()
# print("Minutes taken = " + str((toc-tic)/60.0))

In [591]:
# # Generate the "error rate" vs. "n_estimators" plot.
# plt.figure(figsize=(12, 8), dpi=80)
# for label, clf_err in error_rate.items():
#     xs, ys = zip(*clf_err)
#     plt.plot(xs, ys, label=label, marker='x')

# plt.xlabel("n_estimators")
# plt.ylabel("MSE")
# plt.legend(loc="upper right")
# plt.grid()
# matplotlib.rcParams.update({'font.size': fontsize})
# matplotlib.rcParams['xtick.labelsize'] = ticklabelsize
# matplotlib.rcParams['ytick.labelsize'] = ticklabelsize

In [592]:
# error_rate

In [593]:
# learning_rate_opt, n_estimators_opt, mse_min = get_params(error_rate)
# learning_rate_opt, n_estimators_opt, mse_min

## Plot the error rate versus n_estimators, varying min_child_weight (default = 1)

In [594]:
# ensemble_clfs = [
#     (1,
#         XGBRegressor(seed=model_seed,
#                       max_depth=max_depth_opt,
#                       learning_rate=learning_rate_opt,
#                       min_child_weight=1)),
#     (10,
#         XGBRegressor(seed=model_seed,
#                       max_depth=max_depth_opt,
#                       learning_rate=learning_rate_opt,
#                       min_child_weight=10)),
#      (50,
#         XGBRegressor(seed=model_seed,
#                       max_depth=max_depth_opt,
#                       learning_rate=learning_rate_opt,
#                       min_child_weight=10)),
# ]

# # Map a classifier name to a list of (<n_estimators>, <error rate>) pairs.
# error_rate = OrderedDict((label, []) for label, _ in ensemble_clfs)

# # Range of `n_estimators` values to explore.
# estimators_list = [10]
# estimators_list.extend(range(100, 400, 200))

# tic = time.time()
# for label, clf in ensemble_clfs:
#     print(label)
#     for i in estimators_list:
#         print(i)
#         tic1 = time.time()
#         clf.set_params(n_estimators=i)
#         clf.fit(X_train, y_train['target'])
        
#         pred = clf.predict(X_cv).astype('int')
#         error_rate[label].append((i, mean_squared_error(y_cv['target'], pred)))
#         toc1 = time.time()
#         print("Time taken = " + str((toc1-tic1)/60.0) + " mins")

# toc = time.time()
# print("Minutes taken = " + str((toc-tic)/60.0))

In [595]:
# # Generate the "error rate" vs. "n_estimators" plot.
# plt.figure(figsize=(12, 8), dpi=80)
# for label, clf_err in error_rate.items():
#     xs, ys = zip(*clf_err)
#     plt.plot(xs, ys, label=label, marker='x')

# plt.xlabel("n_estimators")
# plt.ylabel("MSE")
# plt.legend(loc="upper right")
# plt.grid()
# matplotlib.rcParams.update({'font.size': fontsize})
# matplotlib.rcParams['xtick.labelsize'] = ticklabelsize
# matplotlib.rcParams['ytick.labelsize'] = ticklabelsize

In [596]:
# error_rate

In [597]:
# min_child_weight_opt, n_estimators_opt, mse_min = get_params(error_rate)
# min_child_weight_opt, n_estimators_opt, mse_min

In [598]:
n_estimators_opt = 300
max_depth_opt = 9
learning_rate_opt = 0.01
min_child_weight_opt = 1

# Retrain model with optimum hyperparameters

In [599]:
# Create the model
model = XGBRegressor(seed=model_seed,
                      n_estimators=n_estimators_opt,
                      max_depth=max_depth_opt,
                      learning_rate=learning_rate_opt,
                      min_child_weight=min_child_weight_opt)

# # Train the regressor
# model.fit(X_train, y_train['target'])

# # Do prediction on train set
# pred = model.predict(X_train).astype('int')

# # Calculate RMSE
# print("Train set MSE = " + str(mean_squared_error(y_train['target'], pred)))

# # Do prediction on val set
# pred = model.predict(X_cv).astype('int')

# # Calculate RMSE
# print("Val set MSE = " + str(mean_squared_error(y_cv['target'], pred)))

By tuning hyperparams, MSE of the val set dropped from 5.41 to 4.98

# Retrain model on train+val set

In [600]:
# Train the regressor
model.fit(X_train_cv, y_train_cv['target'])

[09:11:26] WARNING: /Users/travis/build/dmlc/xgboost/src/gbm/gbtree.cc:139: Tree method is automatically selected to be 'approx' for faster speed. To use old behavior (exact greedy algorithm on single machine), set tree_method to 'exact'.


XGBRegressor(base_score=0.5, booster='gbtree', colsample_bylevel=1,
             colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
             importance_type='gain', interaction_constraints='',
             learning_rate=0.01, max_delta_step=0, max_depth=9,
             min_child_weight=1, missing=nan, monotone_constraints='()',
             n_estimators=300, n_jobs=0, num_parallel_tree=1,
             objective='reg:squarederror', random_state=100, reg_alpha=0,
             reg_lambda=1, scale_pos_weight=1, seed=100, subsample=1,
             tree_method='approx', validate_parameters=1, verbosity=None)

# Get predictions for validation set 28 days

In [601]:
# Rename columns
sale_tr_val_melt = sale_tr_val_melt.rename(columns={'variable': 'date_block_num', 
                                                    'value': 'target'})

In [602]:
# Convert variable to a numerical column ie. d_1 to 1
sale_tr_val_melt['date_block_num'] = sale_tr_val_melt['date_block_num'].str.split('_').str[1].astype('int16')
sale_tr_val_melt.tail()

id      item_id  dept_id cat_id store_id  \
5832732  FOODS_3_823_WI_3_validation  FOODS_3_823  FOODS_3  FOODS     WI_3   
5832733  FOODS_3_824_WI_3_validation  FOODS_3_824  FOODS_3  FOODS     WI_3   
5832734  FOODS_3_825_WI_3_validation  FOODS_3_825  FOODS_3  FOODS     WI_3   
5832735  FOODS_3_826_WI_3_validation  FOODS_3_826  FOODS_3  FOODS     WI_3   
5832736  FOODS_3_827_WI_3_validation  FOODS_3_827  FOODS_3  FOODS     WI_3   

        state_id  date_block_num  target  
5832732       WI            1913       1  
5832733       WI            1913       0  
5832734       WI            1913       0  
5832735       WI            1913       3  
5832736       WI            1913       0

In [603]:
# Convert variable to a numerical column ie. d_1 to 1
calendar['d'] = calendar['d'].str.split('_').str[1].astype('int16')
calendar.head()

date  wm_yr_wk  wday  month  year  d event_name_1 event_type_1  \
0  2011-01-29     11101     1      1  2011  1          nil          nil   
1  2011-01-30     11101     2      1  2011  2          nil          nil   
2  2011-01-31     11101     3      1  2011  3          nil          nil   
3  2011-02-01     11101     4      2  2011  4          nil          nil   
4  2011-02-02     11101     5      2  2011  5          nil          nil   

  event_name_2 event_type_2  snap_CA  snap_TX  snap_WI  
0          nil          nil        0        0        0  
1          nil          nil        0        0        0  
2          nil          nil        0        0        0  
3          nil          nil        1        1        0  
4          nil          nil        1        0        1

In [604]:
# Do prediction on the test set
results = y_cv[['id']].copy()
sale_tr_val_melt2 = sale_tr_val_melt[sale_tr_val_melt['date_block_num']<pred_start]
mse = []

for i in range(pred_start, pred_end+1):
    print("\nday = " + str(i))
    
    # Create the test set here
    temp = sale_tr_val_melt2[sale_tr_val_melt2['date_block_num']==1].copy()
    temp.loc[:, 'target'] = 0
    temp.loc[:, 'date_block_num'] = i
    sale_tr_val_melt2 = sale_tr_val_melt2.append(temp)
    
    # Merge with calendar to get events info
    sale_tr_val_melt_merged = sale_tr_val_melt2.merge(calendar, 
                                                     left_on='date_block_num', 
                                                     right_on='d', 
                                                     how='left')

    # Drop unneccesary columns
    sale_tr_val_melt_merged.drop(columns=['d'], inplace=True)
    
    # Merge with sell_prices to get price info
    all_data = sale_tr_val_melt_merged.merge(sell_prices, 
                                             left_on=['store_id', 'item_id', 'wm_yr_wk'], 
                                             right_on=['store_id', 'item_id', 'wm_yr_wk'], 
                                             how='left')
    
    # Remove those rows where sell price is NaN and value=0 as this means the item is not being sold yet
    all_data.dropna(axis=0, how='any', subset=['sell_price'], inplace=True)

    # Remove unnecessaray column
    all_data.drop(['date'], axis=1, inplace=True)
    
    # Convert features to categorical 
    for column in columns_to_be_labeled:
        all_data.loc[:, column] = LabelEncoder().fit_transform(all_data[column]) # fit_transform: Fit label encoder and return encoded labels. Note the encoded labels are integers!!

    # Get the lag features
    all_data = get_lags(index_cols, cols_to_rename, all_data, shift_range)
    
    # Remove those rows where target_lag_x is NaN
    all_data.dropna(axis=0, how='any', subset=lag_cols, inplace=True)
    
    # Get the test set
    X_test = all_data[all_data['date_block_num']==i].drop(to_drop_cols, axis = 1)
    y_test = all_data[all_data['date_block_num']==i][['id', 'target']]
    
    # Do prediction on test set
    pred = model.predict(X_test).astype('int')
    print("No. of zeros = " + str(len(pred[pred==0])))
    
    # Add preds to dataset
    sale_tr_val_melt2.loc[sale_tr_val_melt2['date_block_num']==i, 'target'] = pred 

    # Calculate RMSE
    print("MSE = " + str(mean_squared_error(y_test['target'], pred)))
    mse.append(mean_squared_error(y_test['target'], pred))
    
    results['F'+str(i-pred_start+1)] = pred
    
print("mse mean = " + str(np.mean(mse)))
results


day = 1914
No. of zeros = 2301
MSE = 6.4286651361102

day = 1915
No. of zeros = 2385
MSE = 5.019678583142014

day = 1916
No. of zeros = 2437
MSE = 4.375532961626763

day = 1917
No. of zeros = 2471
MSE = 4.467366349622827

day = 1918
No. of zeros = 2475
MSE = 7.12856674319449

day = 1919
No. of zeros = 2459
MSE = 11.591669399803214

day = 1920
No. of zeros = 2480
MSE = 8.950475565759266

day = 1921
No. of zeros = 2561
MSE = 5.620859298130535

day = 1922
No. of zeros = 2607
MSE = 4.934732699245655

day = 1923
No. of zeros = 2678
MSE = 3.461134798294523

day = 1924
No. of zeros = 2675
MSE = 4.604460478845523

day = 1925
No. of zeros = 2672
MSE = 7.672679567071171

day = 1926
No. of zeros = 2673
MSE = 9.619547392587734

day = 1927
No. of zeros = 2669
MSE = 9.208592981305346

day = 1928
No. of zeros = 2694
MSE = 5.323384716300426

day = 1929
No. of zeros = 2724
MSE = 3.6894063627418827

day = 1930
No. of zeros = 2729
MSE = 4.387340111511971

day = 1931
No. of zeros = 2733
MSE = 4.665792062

id  F1  F2  F3  F4  F5  F6  F7  F8  F9  \
4683620  HOBBIES_1_001_WI_3_validation   0   0   0   0   0   0   0   0   0   
4683621  HOBBIES_1_002_WI_3_validation   0   0   0   0   0   0   0   0   0   
4683622  HOBBIES_1_003_WI_3_validation   0   0   0   0   0   0   0   0   0   
4683623  HOBBIES_1_004_WI_3_validation   0   0   0   0   0   0   0   0   0   
4683624  HOBBIES_1_005_WI_3_validation   0   0   0   0   0   0   0   0   0   
4683625  HOBBIES_1_006_WI_3_validation   1   1   1   0   0   0   0   0   0   
4683626  HOBBIES_1_007_WI_3_validation   0   0   0   0   0   0   0   0   0   
4683627  HOBBIES_1_008_WI_3_validation   2   2   2   2   2   2   2   2   2   
4683628  HOBBIES_1_009_WI_3_validation   0   0   0   0   0   0   0   0   0   
4683629  HOBBIES_1_010_WI_3_validation   0   0   0   0   0   0   0   0   0   
4683630  HOBBIES_1_011_WI_3_validation   0   0   0   0   0   0   0   0   0   
4683631  HOBBIES_1_012_WI_3_validation   0   0   0   0   0   0   0   0   0   
4683632  HOBBIES_1_013_WI_3_validation   0   0   0   0   0   0   0   0   0   
4683633  HOBBIES_1_014_WI_3_validation   1   1   1   1   1   2   1   1   1   
4683634  HOBBIES_1_015_WI_3_validation   1   1   1   2   2   2   2   1   1   
4683635  HOBBIES_1_016_WI_3_validation   2   1   1   1   2   2   2   1   1   
4683636  HOBBIES_1_017_WI_3_validation   0   0   0   0   0   0   0   0   0   
4683637  HOBBIES_1_018_WI_3_validation   0   0   0   0   0   0   0   0   0   
4683638  HOBBIES_1_019_WI_3_validation   0   0   0   0   0   1   1   1   1   
4683639  HOBBIES_1_020_WI_3_validation   0   0   0   0   0   0   0   0   0   
4683640  HOBBIES_1_021_WI_3_validation   0   0   0   0   0   0   0   0   0   
4683641  HOBBIES_1_022_WI_3_validation   0   0   0   0   0   0   0   0   0   
4683642  HOBBIES_1_023_WI_3_validation   0   0   0   0   0   0   0   0   0   
4683643  HOBBIES_1_024_WI_3_validation   0   0   0   0   0   0   0   0   0   
4683644  HOBBIES_1_025_WI_3_validation   0   0   0   0   0   0   0   0   0   
4683645  HOBBIES_1_026_WI_3_validation   0   0   0   0   0   0   0   0   0   
4683646  HOBBIES_1_027_WI_3_validation   0   0   0   0   0   0   0   0   0   
4683647  HOBBIES_1_028_WI_3_validation   0   0   0   0   0   0   0   0   0   
4683648  HOBBIES_1_029_WI_3_validation   0   0   0   0   0   0   0   0   0   
4683649  HOBBIES_1_030_WI_3_validation   1   0   1   1   1   1   1   2   2   
...                                ...  ..  ..  ..  ..  ..  ..  ..  ..  ..   
4686639    FOODS_3_798_WI_3_validation   0   0   0   0   0   0   0   0   0   
4686640    FOODS_3_799_WI_3_validation   0   0   0   0   0   0   0   0   0   
4686641    FOODS_3_800_WI_3_validation   6   5   5   4   5   6   5   5   5   
4686642    FOODS_3_801_WI_3_validation   2   2   2   1   2   3   2   2   2   
4686643    FOODS_3_802_WI_3_validation   0   0   0   0   0   0   0   0   0   
4686644    FOODS_3_803_WI_3_validation   0   0   0   0   0   0   0   0   0   
4686645    FOODS_3_804_WI_3_validation   3   4   4   2   2   3   3   4   3   
4686646    FOODS_3_805_WI_3_validation   1   1   1   1   1   1   1   1   1   
4686647    FOODS_3_806_WI_3_validation   0   0   0   0   0   0   0   0   0   
4686648    FOODS_3_807_WI_3_validation   1   1   1   1   1   1   0   1   1   
4686649    FOODS_3_808_WI_3_validation   0   0   0   0   0   0   0   0   0   
4686650    FOODS_3_809_WI_3_validation   1   1   1   1   1   1   1   1   1   
4686651    FOODS_3_810_WI_3_validation   0   0   0   0   0   0   0   0   0   
4686652    FOODS_3_811_WI_3_validation  18  17  16  15  15  16  20  17  15   
4686653    FOODS_3_812_WI_3_validation   0   0   0   0   0   0   0   0   0   
4686654    FOODS_3_813_WI_3_validation   0   0   0   0   0   0   0   0   0   
4686655    FOODS_3_814_WI_3_validation   1   1   1   1   1   2   1   1   1   
4686656    FOODS_3_815_WI_3_validation   0   0   0   0   0   0   0   0   0   
4686657    FOODS_3_816_WI_3_validation   1   2   2   3   3   3   3   4   5   
4686658    FOODS_3_817_WI_3_validation   0   0   0   0   0

In [605]:
# Store results as csv
results.to_csv("./out/v6_results-" + state + "-" + store + "-" + str(pred_start) + "-" + str(pred_end) + ".csv", index=False)

In [606]:
tocEnd = time.time()
print("Total time = " + str((tocEnd-ticStart)/60.0) + " mins")

Total time = 48.31221963564555 mins


# Combine all predictions together and prepare submission file

In [609]:
# Combine all predictions together
pred_start = 1914              # do pred starting from this day
pred_end = 1941                # do pred until and include this day

state_store_list = [('CA', 'CA_1'), ('CA', 'CA_2'), ('CA', 'CA_3'), ('CA', 'CA_4'),
                    ('TX', 'TX_1'), ('TX', 'TX_2'), ('TX', 'TX_3'), 
                    ('WI', 'WI_1'), ('WI', 'WI_2'), ('WI', 'WI_3')]

state_store = state_store_list[0]
results_tot = pd.read_csv("./out/v5_results-" + state_store[0] + "-" + state_store[1] + "-" + str(pred_start) + "-" + str(pred_end) + ".csv")


for state_store in state_store_list[1:]:
    results = pd.read_csv("./out/v6_results-" + state_store[0] + "-" + state_store[1] + "-" + str(pred_start) + "-" + str(pred_end) + ".csv")
    results_tot = results_tot.append(results)
    
results_tot

id  F1  F2  F3  F4  F5  F6  F7  F8  F9  ...  \
0     HOBBIES_1_001_CA_1_validation   0   0   0   0   0   0   0   0   0  ...   
1     HOBBIES_1_002_CA_1_validation   0   0   0   0   0   0   0   0   0  ...   
2     HOBBIES_1_003_CA_1_validation   0   0   0   0   0   0   0   0   0  ...   
3     HOBBIES_1_004_CA_1_validation   2   1   1   1   1   2   2   1   0  ...   
4     HOBBIES_1_005_CA_1_validation   1   1   1   1   1   2   1   1   1  ...   
5     HOBBIES_1_006_CA_1_validation   0   0   0   0   0   0   0   0   0  ...   
6     HOBBIES_1_007_CA_1_validation   0   0   0   0   0   0   0   0   0  ...   
7     HOBBIES_1_008_CA_1_validation   6   5   6   6   8   7   9   9   8  ...   
8     HOBBIES_1_009_CA_1_validation   0   0   0   0   0   0   0   0   0  ...   
9     HOBBIES_1_010_CA_1_validation   0   0   0   0   0   0   0   0   0  ...   
10    HOBBIES_1_011_CA_1_validation   0   0   0   0   0   0   0   0   0  ...   
11    HOBBIES_1_012_CA_1_validation   0   0   0   0   0   0   0   0   0  ...   
12    HOBBIES_1_013_CA_1_validation   0   0   0   0   0   0   0   0   0  ...   
13    HOBBIES_1_014_CA_1_validation   1   1   1   1   1   1   1   0   0  ...   
14    HOBBIES_1_015_CA_1_validation   2   3   3   3   3   4   5   3   3  ...   
15    HOBBIES_1_016_CA_1_validation   5   4   4   5   7   6   7   5   5  ...   
16    HOBBIES_1_017_CA_1_validation   1   0   0   0   1   1   1   0   0  ...   
17    HOBBIES_1_018_CA_1_validation   0   0   0   0   0   0   0   0   0  ...   
18    HOBBIES_1_019_CA_1_validation   6   6   7   8   8  10  10   8   9  ...   
19    HOBBIES_1_020_CA_1_validation   0   0   0   0   0   0   0   0   0  ...   
20    HOBBIES_1_021_CA_1_validation   0   0   0   0   0   0   0   0   0  ...   
21    HOBBIES_1_022_CA_1_validation   0   0   0   0   0   0   0   0   0  ...   
22    HOBBIES_1_023_CA_1_validation   1   1   0   1   0   1   1   0   0  ...   
23    HOBBIES_1_024_CA_1_validation   0   0   0   0   0   0   0   0   0  ...   
24    HOBBIES_1_025_CA_1_validation   0   0   0   0   0   0   0   0   0  ...   
25    HOBBIES_1_026_CA_1_validation   0   0   0   0   0   0   0   0   0  ...   
26    HOBBIES_1_027_CA_1_validation   0   0   0   0   0   0   0   0   0  ...   
27    HOBBIES_1_028_CA_1_validation   0   0   0   0   0   0   0   0   0  ...   
28    HOBBIES_1_029_CA_1_validation   1   1   1   1   0   1   1   1   0  ...   
29    HOBBIES_1_030_CA_1_validation   5   7   6   7   7   9  10   6   6  ...   
...                             ...  ..  ..  ..  ..  ..  ..  ..  ..  ..  ...   
3019    FOODS_3_798_WI_3_validation   0   0   0   0   0   0   0   0   0  ...   
3020    FOODS_3_799_WI_3_validation   0   0   0   0   0   0   0   0   0  ...   
3021    FOODS_3_800_WI_3_validation   6   5   5   4   5   6   5   5   5  ...   
3022    FOODS_3_801_WI_3_validation   2   2   2   1   2   3   2   2   2  ...   
3023    FOODS_3_802_WI_3_validation   0   0   0   0   0   0   0   0   0  ...   
3024    FOODS_3_803_WI_3_validation   0   0   0   0   0   0   0   0   0  ...   
3025    FOODS_3_804_WI_3_validation   3   4   4   2   2   3   3   4   3  ...   
3026    FOODS_3_805_WI_3_validation   1   1   1   1   1   1   1   1   1  ...   
3027    FOODS_3_806_WI_3_validation   0   0   0   0   0   0   0   0   0  ...   
3028    FOODS_3_807_WI_3_validation   1   1   1   1   1   1   0   1   1  ...   
3029    FOODS_3_808_WI_3_validation   0   0   0   0   0   0   0   0   0  ...   
3030    FOODS_3_809_WI_3_validation   1   1   1   1   1   1   1   1   1  ...   
3031    FOODS_3_810_WI_3_validation   0   0   0   0   0   0   0   0   0  ...   
3032    FOODS_3_811_WI_3_validation  18  17  16  15  15  16  20  17  15  ...   
3033    FOODS_3_812_WI_3_validation   0   0   0   0   0   0   0   0   0  ...   
3034    FOODS_3_813_WI_3_validation   0   0   0   0   0   0   0   0   0  ...   
3035    FOODS_3_814_WI_3_validation   1   1   1   1   1   2   1   1   1  ...   
3036    FOODS_3_815_WI_3_validation   0   0   0   0   0   0   0   0   0  ...   
3037    FOODS_3_816_WI_3_validati

In [610]:
# Prepare submission file
submission = pd.merge(sample_sub[['id']], results_tot, on='id', how='left')
submission.fillna(value=0, inplace=True)
submission.to_csv("./out/v6_submission.csv", index=False)
submission

id   F1   F2   F3   F4   F5    F6    F7  \
0      HOBBIES_1_001_CA_1_validation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
1      HOBBIES_1_002_CA_1_validation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
2      HOBBIES_1_003_CA_1_validation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
3      HOBBIES_1_004_CA_1_validation  2.0  1.0  1.0  1.0  1.0   2.0   2.0   
4      HOBBIES_1_005_CA_1_validation  1.0  1.0  1.0  1.0  1.0   2.0   1.0   
5      HOBBIES_1_006_CA_1_validation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
6      HOBBIES_1_007_CA_1_validation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
7      HOBBIES_1_008_CA_1_validation  6.0  5.0  6.0  6.0  8.0   7.0   9.0   
8      HOBBIES_1_009_CA_1_validation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
9      HOBBIES_1_010_CA_1_validation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
10     HOBBIES_1_011_CA_1_validation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
11     HOBBIES_1_012_CA_1_validation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
12     HOBBIES_1_013_CA_1_validation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
13     HOBBIES_1_014_CA_1_validation  1.0  1.0  1.0  1.0  1.0   1.0   1.0   
14     HOBBIES_1_015_CA_1_validation  2.0  3.0  3.0  3.0  3.0   4.0   5.0   
15     HOBBIES_1_016_CA_1_validation  5.0  4.0  4.0  5.0  7.0   6.0   7.0   
16     HOBBIES_1_017_CA_1_validation  1.0  0.0  0.0  0.0  1.0   1.0   1.0   
17     HOBBIES_1_018_CA_1_validation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
18     HOBBIES_1_019_CA_1_validation  6.0  6.0  7.0  8.0  8.0  10.0  10.0   
19     HOBBIES_1_020_CA_1_validation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
20     HOBBIES_1_021_CA_1_validation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
21     HOBBIES_1_022_CA_1_validation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
22     HOBBIES_1_023_CA_1_validation  1.0  1.0  0.0  1.0  0.0   1.0   1.0   
23     HOBBIES_1_024_CA_1_validation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
24     HOBBIES_1_025_CA_1_validation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
25     HOBBIES_1_026_CA_1_validation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
26     HOBBIES_1_027_CA_1_validation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
27     HOBBIES_1_028_CA_1_validation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
28     HOBBIES_1_029_CA_1_validation  1.0  1.0  1.0  1.0  0.0   1.0   1.0   
29     HOBBIES_1_030_CA_1_validation  5.0  7.0  6.0  7.0  7.0   9.0  10.0   
...                              ...  ...  ...  ...  ...  ...   ...   ...   
60950    FOODS_3_798_WI_3_evaluation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
60951    FOODS_3_799_WI_3_evaluation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
60952    FOODS_3_800_WI_3_evaluation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
60953    FOODS_3_801_WI_3_evaluation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
60954    FOODS_3_802_WI_3_evaluation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
60955    FOODS_3_803_WI_3_evaluation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
60956    FOODS_3_804_WI_3_evaluation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
60957    FOODS_3_805_WI_3_evaluation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
60958    FOODS_3_806_WI_3_evaluation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
60959    FOODS_3_807_WI_3_evaluation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
60960    FOODS_3_808_WI_3_evaluation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
60961    FOODS_3_809_WI_3_evaluation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
60962    FOODS_3_810_WI_3_evaluation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
60963    FOODS_3_811_WI_3_evaluation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
60964    FOODS_3_812_WI_3_evaluation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
60965    FOODS_3_813_WI_3_evaluation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
60966    FOODS_3_814_WI_3_evaluation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
60967    FOODS_3_815_WI_3_evaluation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
60968    FOODS_3_816_WI_3_evaluation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
60969    FOODS_3_817_WI_3_evaluation  0.0  0.0  0.0  0.0  0.0   0.0   0.0   
60970    FOODS_3_818_WI_3_eval

This obtained weighted root mean squared scaled error (RMSSE) of 3.53072 on the validation set

In [616]:
pd.set_option('display.max_columns', 500)
sales_tr_val = pd.read_csv("./data/sales_train_validation.csv")
print("len(sales_tr_val) = " + str(len(sales_tr_val)))
sales_tr_val

len(sales_tr_val) = 30490


id        item_id    dept_id   cat_id  \
0      HOBBIES_1_001_CA_1_validation  HOBBIES_1_001  HOBBIES_1  HOBBIES   
1      HOBBIES_1_002_CA_1_validation  HOBBIES_1_002  HOBBIES_1  HOBBIES   
2      HOBBIES_1_003_CA_1_validation  HOBBIES_1_003  HOBBIES_1  HOBBIES   
3      HOBBIES_1_004_CA_1_validation  HOBBIES_1_004  HOBBIES_1  HOBBIES   
4      HOBBIES_1_005_CA_1_validation  HOBBIES_1_005  HOBBIES_1  HOBBIES   
5      HOBBIES_1_006_CA_1_validation  HOBBIES_1_006  HOBBIES_1  HOBBIES   
6      HOBBIES_1_007_CA_1_validation  HOBBIES_1_007  HOBBIES_1  HOBBIES   
7      HOBBIES_1_008_CA_1_validation  HOBBIES_1_008  HOBBIES_1  HOBBIES   
8      HOBBIES_1_009_CA_1_validation  HOBBIES_1_009  HOBBIES_1  HOBBIES   
9      HOBBIES_1_010_CA_1_validation  HOBBIES_1_010  HOBBIES_1  HOBBIES   
10     HOBBIES_1_011_CA_1_validation  HOBBIES_1_011  HOBBIES_1  HOBBIES   
11     HOBBIES_1_012_CA_1_validation  HOBBIES_1_012  HOBBIES_1  HOBBIES   
12     HOBBIES_1_013_CA_1_validation  HOBBIES_1_013  HOBBIES_1  HOBBIES   
13     HOBBIES_1_014_CA_1_validation  HOBBIES_1_014  HOBBIES_1  HOBBIES   
14     HOBBIES_1_015_CA_1_validation  HOBBIES_1_015  HOBBIES_1  HOBBIES   
15     HOBBIES_1_016_CA_1_validation  HOBBIES_1_016  HOBBIES_1  HOBBIES   
16     HOBBIES_1_017_CA_1_validation  HOBBIES_1_017  HOBBIES_1  HOBBIES   
17     HOBBIES_1_018_CA_1_validation  HOBBIES_1_018  HOBBIES_1  HOBBIES   
18     HOBBIES_1_019_CA_1_validation  HOBBIES_1_019  HOBBIES_1  HOBBIES   
19     HOBBIES_1_020_CA_1_validation  HOBBIES_1_020  HOBBIES_1  HOBBIES   
20     HOBBIES_1_021_CA_1_validation  HOBBIES_1_021  HOBBIES_1  HOBBIES   
21     HOBBIES_1_022_CA_1_validation  HOBBIES_1_022  HOBBIES_1  HOBBIES   
22     HOBBIES_1_023_CA_1_validation  HOBBIES_1_023  HOBBIES_1  HOBBIES   
23     HOBBIES_1_024_CA_1_validation  HOBBIES_1_024  HOBBIES_1  HOBBIES   
24     HOBBIES_1_025_CA_1_validation  HOBBIES_1_025  HOBBIES_1  HOBBIES   
25     HOBBIES_1_026_CA_1_validation  HOBBIES_1_026  HOBBIES_1  HOBBIES   
26     HOBBIES_1_027_CA_1_validation  HOBBIES_1_027  HOBBIES_1  HOBBIES   
27     HOBBIES_1_028_CA_1_validation  HOBBIES_1_028  HOBBIES_1  HOBBIES   
28     HOBBIES_1_029_CA_1_validation  HOBBIES_1_029  HOBBIES_1  HOBBIES   
29     HOBBIES_1_030_CA_1_validation  HOBBIES_1_030  HOBBIES_1  HOBBIES   
...                              ...            ...        ...      ...   
30460    FOODS_3_798_WI_3_validation    FOODS_3_798    FOODS_3    FOODS   
30461    FOODS_3_799_WI_3_validation    FOODS_3_799    FOODS_3    FOODS   
30462    FOODS_3_800_WI_3_validation    FOODS_3_800    FOODS_3    FOODS   
30463    FOODS_3_801_WI_3_validation    FOODS_3_801    FOODS_3    FOODS   
30464    FOODS_3_802_WI_3_validation    FOODS_3_802    FOODS_3    FOODS   
30465    FOODS_3_803_WI_3_validation    FOODS_3_803    FOODS_3    FOODS   
30466    FOODS_3_804_WI_3_validation    FOODS_3_804    FOODS_3    FOODS   
30467    FOODS_3_805_WI_3_validation    FOODS_3_805    FOODS_3    FOODS   
30468    FOODS_3_806_WI_3_validation    FOODS_3_806    FOODS_3    FOODS   
30469    FOODS_3_807_WI_3_validation    FOODS_3_807    FOODS_3    FOODS   
30470    FOODS_3_808_WI_3_validation    FOODS_3_808    FOODS_3    FOODS   
30471    FOODS_3_809_WI_3_validation    FOODS_3_809    FOODS_3    FOODS   
30472    FOODS_3_810_WI_3_validation    FOODS_3_810    FOODS_3    FOODS   
30473    FOODS_3_811_WI_3_validation    FOODS_3_811    FOODS_3    FOODS   
30474    FOODS_3_812_WI_3_validation    FOODS_3_812    FOODS_3    FOODS   
30475    FOODS_3_813_WI_3_validation    FOODS_3_813    FOODS_3    FOODS   
30476    FOODS_3_814_WI_3_validation    FOODS_3_814    FOODS_3    FOODS   
30477    FOODS_3_815_WI_3_validation    FOODS_3_815    FOODS_3    FOODS   
30478    FOODS_3_816_WI_3_validation    FOODS_3_816    FOODS_3    FOODS   
30479    FOODS_3_817_WI_3_validation    FOODS_3_817    FOODS_3    FOODS   
30480    FOODS_3_818_WI_3_validation    FOODS_3_818    FOODS_3    FOODS   
30481    FOODS_3_819_WI_3_validation    FOODS_3_819    FOOD